### ISSUE TITLE
New referral type for opt-in/leads (and possibly others)

### ISSUE THREAD

**Issue by amdrew** [AffiliateWP/AffiliateWP#2198]
In our meeting we discussed generating referrals when a user "opts-in". This would generally occur when the visitor (aka lead) signs up to a mailing list. Rather than building this as an add-on (https://trello.com/c/XsqTj2PL/244-opt-in-referrals) we're exploring the possibility of including this in core since we already create referrals when a form is submitted, when a purchase is made etc.

For the above to happen, we should allow custom referral types to be created. This would also allow us to filter referrals and see _how_ referrals were generated. Was the referral generated from an eCommerce purchase? From a form submission? From an opt-in?

Here's another example of how custom referral types would help other developers:
https://github.com/AffiliateWP/AffiliateWP/issues/1100

An older related discussion:
https://github.com/AffiliateWP/AffiliateWP/issues/505

**Comment by pippinsplugins** [comment]
When we do this, we can also then add integrations for plugins like these:

- OptinMonster
- MailChimp for WordPress
- Popup Builder
- Etc

**Comment by DrewAPicture** [comment]
Changed "status/statuses" to "type/types". Statuses are statuses, types are types ;)

**Comment by pippinsplugins** [comment]
I've added a new `type` column and set all existing referrals to the type of `sale` through an upgraded routine.

For the moment I have decided against requiring that referral types be formally registered (unlike post types in WP core). The primary reason for this is that we're not (at least for now) revealing anything through the UI for these referral types, so it's not necessary that the available types be strictly known.

**Comment by pippinsplugins** [comment]
I've changed my mind on having a defined method for registering referral types.

We need to have a way to filter the Referrals table by type so that depends on having a registry where we can retrieve a list of all available types.

I've added a registry and added a filter dropdown to the referrals table.

**Comment by pippinsplugins** [comment]
This is ready for testing.

By default we're now registering two core types:

- `sale`
- `opt-in`

New types can be registered like this:

```php
function pw_affwp_register_referral_type( $registry ) {
	$registry->register_type( 'ping', array( 'label' => 'Ping' ) );
}
add_action('affwp_referral_type_init', 'pw_affwp_register_referral_type' );
```

The `type` parameter is supported in the `add()`, `update()`, and `get_referrals()` methods.

The Referrals list table can be filtered by any registered type.

If a referral is created with a non-registered type, it will default to `sale`.

**Comment by amdrew** [comment]
@pippinsplugins We should add a description on the edit referral screen for `Type` and also add a note on how it cannot be edited. Something like:

`The referral type. This cannot be changed.`

**Comment by amdrew** [comment]
@pippinsplugins I wonder if we should hide the types select menu on the referrals screen if only 1 type exists on the site. Ideally it should also only show the referral types that exist on the site.

On my test site there are only `Sale` referrals and could be confusing to some if they can filter by `Opt-In` but no opt-in referrals actually exist.

**Comment by amdrew** [comment]
@pippinsplugins The new opt-in referral type select menu on a Gravity Form is echoing the selected state (and a random full stop) to the top of the screen when selected and saved.

<img width="474" alt="screenshot 2018-01-16 10 51 32" src="https://user-images.githubusercontent.com/52581/34963058-5b605460-faab-11e7-8468-b24f53057200.png">

**Comment by pippinsplugins** [comment]
@amdrew description added and selected state output fixed.

I don't think we should hide the type dropdown for the same reason that we don't hide the status links when no referrals of those statuses are present. The presence of the UI helps educate the site owner on what options are available, even if those options are not currently used.

**Comment by amdrew** [comment]
@pippinsplugins We should consider allowing the referral type to be changed from the edit referral screen.

**Comment by pippinsplugins** [comment]
Referral types can now be changed.

**Comment by pippinsplugins** [comment]
Fixed the problem with `opt-in` types not saving.

**Comment by amdrew** [comment]
@pippinsplugins 

Getting these notices outputted when testing Gravity Forms:

```
[15-Jan-2018 22:29:38 UTC] WordPress database error You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'NULL1' at line 1 for query UPDATE wp_22_affiliate_wp_referrals SET type = 'sale' where type IS NULL1; made by do_action('admin_init'), WP_Hook->do_action, WP_Hook->apply_filters, call_user_func_array, Affiliate_WP_Upgrades->init, Affiliate_WP_Upgrades->v22_upgrade
```

**Comment by pippinsplugins** [comment]
Fixed.

**Comment by amdrew** [comment]
After discussing this with @pippinsplugins, we're also going to add a `lead` referral type. Form integrations would need to support this. 

We're also going to add the `type` to the conversion script shortcode.

**Comment by pippinsplugins** [comment]
Lead type is now registered.

**Comment by pippinsplugins** [comment]
Form integrations to add referral type selection to:

- [ ] Gravity Forms
- [ ] Ninja Forms
- [ ] Formidable Forms
- [ ] CalderaForms
- [ ] Contact Form 7

**Comment by pippinsplugins** [comment]
All form integrations now support referral type selections.

**Comment by pippinsplugins** [comment]
Note: unit tests are failing now due to changes from #958. Will resolves those once it is merge.

**Comment by pippinsplugins** [comment]
This is now ready for solid testing @gspice @michaelbeil!

A few notes:

1. The Referrals list table now displays the referral type as `Sale`, `Opt-In`, or `Lead`.
2. A drop down has been added to the Referrals table for filtering the table by specific referral types.
3. Each of the form integrations mentioned [above](https://github.com/AffiliateWP/AffiliateWP/issues/2198#issuecomment-357841029) now include an option to select the referral type.
4. The `[affiliate_conversion_script]` shortcode now supports a `type` parameter that can be `sale`, `lead`, or `opt-in`.
5. New referral types can be programmatically registered. See [comment above](https://github.com/AffiliateWP/AffiliateWP/issues/2198#issuecomment-357494214).

Once this is complete, it will enable us to take this further with #2511 and more.

**Comment by gspice** [comment]
Did my first test using Gravity Forms.  Tested sale, opt-in and lead. The referral was created with the new and correct type chosen, but received a warning:

<img width="781" alt="lead" src="https://user-images.githubusercontent.com/3018245/35073335-410146cc-fb9d-11e7-9a51-587eecd45235.png">

**Comment by amdrew** [comment]
That's an error related to email stuff I'm working on in another branch. I think it got into this branch somehow. You can safely ignore that.

**Comment by pippinsplugins** [comment]
@amdrew could you please ping me when that branch is done so I can get it brought in here?

**Comment by amdrew** [comment]
@pippinsplugins Yep

**Comment by amdrew** [comment]
@pippinsplugins Ready to roll, see https://github.com/AffiliateWP/AffiliateWP/issues/948#issuecomment-359324257

**Comment by pippinsplugins** [comment]
Thanks. I'll wait til we merge that and then I'll pull it in.

**Comment by michaelbeil** [comment]
Did a test using Gravity Forms and a new referral was generated for the new referral types: sale, opt-in, and lead. I received the following warning:

`Warning: array_key_exists() expects parameter 2 to be array, boolean given in /app/public/wp-content/plugins/AffiliateWP/includes/emails/actions.php on line 240`

**Comment by amdrew** [comment]
@michaelbeil Yep that's correct

**Comment by michaelbeil** [comment]
Roger that.

**Comment by pippinsplugins** [comment]
Branch updated from `release/2.2`.

**Comment by pippinsplugins** [comment]
@michaelbeil @gspice this is ready for new testing!

**Comment by gspice** [comment]
@pippinsplugins If I download the 2198 branch, install and attempt to activate it gives me a fatal error:

`Plugin could not be activated because it triggered a fatal error.`

```
( ! ) Fatal error: Call to a member function get() on null in /app/public/wp-content/plugins/AffiliateWP-issue-2198/includes/install.php on line 36
--
1 | 0.0000 | 247984 | {main}( ) | .../plugins.php:0
2 | 0.7201 | 6124640 | do_action( ) | .../plugins.php:174
3 | 0.7201 | 6125368 | WP_Hook->do_action( ) | .../plugin.php:453
4 | 0.7201 | 6125464 | WP_Hook->apply_filters( ) | .../class-wp-hook.php:310
5 | 0.7201 | 6126616 | call_user_func_array:{/app/public/wp-includes/class-wp-hook.php:286} ( ) | .../class-wp-hook.php:286
6 | 0.7201 | 6127920 | affiliate_wp_install( ) | .../class-wp-hook.php:286
```

**Comment by pippinsplugins** [comment]
@gspice that doesn't look to be related to this branch as that line wasn't changed here.

Could you rename the plugin folder from `AffiliateWP-issue-2198` to `affiliate-wp` and try again?

**Comment by gspice** [comment]
@pippinsplugins Renaming the folder worked!  I was able to activate.

Question - does this have the fix in it for the `Upgrade Database` process, where it handles the case of the Woo order missing that a referral is referencing?   I'm getting this error and confirmed I don't have a WC order `1995` any more:

```
[11-Apr-2018 22:57:41 UTC] PHP Fatal error:  Uncaught Exception: Invalid order. in /app/public/wp-content/plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php:102
Stack trace:
#0 /app/public/wp-content/plugins/woocommerce/includes/class-wc-data-store.php(149): Abstract_WC_Order_Data_Store_CPT->read(Object(WC_Order))
#1 /app/public/wp-content/plugins/woocommerce/includes/abstracts/abstract-wc-order.php(109): WC_Data_Store->read(Object(WC_Order))
#2 /app/public/wp-content/plugins/AffiliateWP/includes/integrations/class-woocommerce.php(365): WC_Abstract_Order->__construct('1995')
#3 /app/public/wp-content/plugins/AffiliateWP/includes/admin/tools/upgrades/class-batch-upgrade-create-customers.php(117): Affiliate_WP_WooCommerce->get_customer('1995')
#4 /app/public/wp-content/plugins/AffiliateWP/includes/admin/ajax-actions.php(180): AffWP\Utils\Batch_Process\Upgrade_Create_Customers->process_step()
#5 /app/public/wp-includes/class-wp-hook.php(286): affwp_process_batch_request('')
#6 /app/public/wp-includes/class-wp-hook.php(310): WP_Ho in /app/public/wp-content/plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php on line 102
```

**Comment by pippinsplugins** [comment]
@gspice looks like I still need to account for that.

**Comment by pippinsplugins** [comment]
@gspice I think I've fixed it. Could you try it one more time?

**Comment by gspice** [comment]
@pippinsplugins Success!  Clicked to upgrade database and it completed, no errors in php log.

**Comment by pippinsplugins** [comment]
Yay!

**Comment by pippinsplugins** [comment]
This has been merged into `release/2.2.`

**Comment by michaelbeil** [comment]
Excellent!
